zerver/lib/webhooks/common.py

```diff
@@ -114,3 +130,104 @@ def validate_extract_webhook_http_header(request: HttpRequest, header: str,
         raise MissingHTTPEventHeader(header)
 
     return extracted_header
+
+class ThirdPartyAPIAmbassador:
+    """ Using an instance of this class, integrations/webhook view code can systematically
+    authenticate with a third party API and then send multiple requests to it, store the results,
+    as well as set triggers to be run before and after the calls. The http_api_callback method of
+    this class is the method that integrations/webhook view code should use to actually call the
+    third party api endpoint. """
+
+    def __init__(self,
+                 bot: UserProfile,
+                 request_kwargs: Dict[str, Any]={},
+                 authentication_handler: Optional[Callable[["ThirdPartyAPIAmbassador"], None]]=None,
+                 request_preprocessor: Optional[Callable[["ThirdPartyAPIAmbassador"], None]]=None,
+                 request_postprocessor: Optional[Callable[["ThirdPartyAPIAmbassador"], None]]=None,
+                 ) -> None:
+        """
+            bot:
+                A UserProfile instance which is a bot. The main reason we need to store this is
+                that we may want to read from or update the BotConfigData of the bot which is
+                acting as the ambassador from Zulip -> ThirdPartyAPI.
+
+            request_kwargs:
+                Extra request information such as headers, cookies, json, files, etc. which will
+                be passed to http_api_callback and thus to the request method from the
+                python-requests library.
+
+            authentication_handler:
+                A single-use function that will be called by passing this ThirdPartyAPIAmbassador
+                instance to it during initialization. There are some ready-made methods in the
+                Zulip library (in this same file) which handle basic schemes like sending an API
+                token with every request as a "GET" parameter.
+
+            request_preprocessor:
+                A function that should take this ThirdPartyAPIAmbassador instance and then do
+                some preprocessing, such as setting updating request_kwargs based on the previous
+                response from calling http_api_callback, starting a log entry, etc.
+                Argument(s) passed during the call: self
+
+            request_postprocessor:
+                Similar to the request_preprocessor, this should be a function that takes this
+                ThirdPartyAPIAmbassador object and then does some postprocessing such as logging
+                the results, triggering another subsequent call, PMing the owner of the bot, etc.
+                Argument(s) passed during the call: self
+
+            response_log: (internal use)
+                A list of all of the responses from calling http_api_callback (in chronological
+                order with index 0 being the oldest response and index -1 being the most recent).
+        """
```

Comment:
I don't think this docstring is super helpful in the code itself. We should properly document this system in our `/api` docs or maybe ReadTheDocs once we merge this PR. I'd recommend removing this for now. :)

Reply:
Stuff like this is documented much better in a tutorial or how-to form, in my opinion. Instead of writing docstrings like this, I would recommend adding a commit that actually uses this code. That willl probably help us figure out how and when we should move forward.

Reply:
Got it. For the time being, I'll remove all of the docstrings and then later add a commit for documentation (which can be squashed with the rest).

>  I would recommend adding a commit that actually uses this code.

As mentioned above, I'll be adding a commit like this for Phabricator soon.